Agent auto-restart fails after timeout — session stuck showing error

## Bug Description

When an agent session times out after 120s of being unresponsive, the UI displays "Agent unresponsive — session will restart automatically" but the session never actually restarts. The user is left stuck with the error message and no way to continue.

## Steps to Reproduce

1. Start an agent session
2. Wait for the agent to become unresponsive (e.g., during a long tool call)
3. After 120s, the prompt times out
4. Observe the red "Agent unresponsive — session will restart automatically" banner

## Expected Behavior

The session should automatically restart and allow the user to continue working.

## Actual Behavior

The session is killed and removed from the session map, but the retry/restart logic fails because the session no longer exists. Console shows:

```
[WARN] [ACP] Prompt timed out after 120s — agent unresponsive
[AcpService] Received event on acp://error: {error: "Agent unresponsive — session will restart automatically", ...}
[INFO] [ACP] Session force-stopped — exiting worker to kill agent process
[INFO] [ACP] Session de41ef8a-... removed from map after worker exit
[ERROR] [AcpStore] Retry failed: "Agent unresponsive — session will restart automatically"
```

The root issue: the session is removed from the backend map, but the frontend sendPrompt retry logic tries to use the now-dead session. The retry fails because the session no longer exists, and no new session is spawned.

## Environment

- macOS (Apple Silicon)
- Seren Desktop latest build
- Multiple agent tabs open (Claude #1, #2, #3)

--- a/src-tauri/src/acp.rs
+++ b/src-tauri/src/acp.rs
@@ -402,6 +402,8 @@ struct ClientDelegate {
     pending_diff_proposals: Arc<Mutex<HashMap<String, oneshot::Sender<bool>>>>,
     /// Maps terminal IDs to tool call IDs for tracking command activity in the UI
     terminal_tool_calls: Arc<Mutex<HashMap<String, String>>>,
+    /// Signalled on every session notification so the prompt timeout resets on activity
+    activity_notify: Arc<tokio::sync::Notify>,
 }
 
 impl ClientDelegate {
@@ -840,6 +842,7 @@ impl Client for ClientDelegate {
     }
 
     async fn session_notification(&self, args: SessionNotification) -> AcpResult<()> {
+        self.activity_notify.notify_one();
         handle_session_notification(&self.app, &self.session_id, args);
         Ok(())
     }
@@ -1398,6 +1401,7 @@ async fn run_session_worker(
 
     // Create client delegate (keep terminals ref for cleanup on session exit)
     let terminals = Arc::new(Mutex::new(crate::terminal::TerminalManager::new()));
+    let activity_notify = Arc::new(tokio::sync::Notify::new());
     let delegate = ClientDelegate {
         app: app.clone(),
         session_id: session_id.clone(),
@@ -1407,6 +1411,7 @@ async fn run_session_worker(
         pending_permissions,
         pending_diff_proposals,
         terminal_tool_calls: Arc::new(Mutex::new(HashMap::new())),
+        activity_notify: Arc::clone(&activity_notify),
     };
 
     // Create ACP connection - use spawn_local since we're in a LocalSet
@@ -1641,8 +1646,11 @@ async fn run_session_worker(
                 let mut cancelled = false;
                 let mut cancel_deadline: Option<tokio::time::Instant> = None;
                 let mut force_stopped = false;
-                let prompt_timeout = tokio::time::sleep(std::time::Duration::from_secs(PROMPT_TIMEOUT_SECS));
-                tokio::pin!(prompt_timeout);
+                // Activity-aware timeout: resets whenever the agent sends a
+                // notification (message chunk, tool call, tool result, etc.).
+                // Only fires when the agent is truly silent for PROMPT_TIMEOUT_SECS.
+                let mut idle_deadline = tokio::time::Instant::now()
+                    + std::time::Duration::from_secs(PROMPT_TIMEOUT_SECS);
 
                 let prompt_result = loop {
                     // If cancel was sent, enforce a deadline for the prompt to finish
@@ -1681,9 +1689,9 @@ async fn run_session_worker(
                             result = &mut prompt_fut => {
                                 break result;
                             }
-                            _ = &mut prompt_timeout => {
+                            _ = tokio::time::sleep_until(idle_deadline) => {
                                 log::warn!(
-                                    "[ACP] Prompt timed out after {}s — agent unresponsive",
+                                    "[ACP] Prompt timed out after {}s of inactivity — agent unresponsive",
                                     PROMPT_TIMEOUT_SECS
                                 );
                                 force_stopped = true;
@@ -1692,6 +1700,11 @@ async fn run_session_worker(
                                         "Agent unresponsive — session will restart automatically".into(),
                                     ),
                                 ));
+                            }
+                            _ = activity_notify.notified() => {
+                                // Agent sent a notification — reset the idle deadline
+                                idle_deadline = tokio::time::Instant::now()
+                                    + std::time::Duration::from_secs(PROMPT_TIMEOUT_SECS);
                             }
                             cmd = command_rx.recv() => {
                                 match cmd {

--- a/src/stores/acp.store.ts
+++ b/src/stores/acp.store.ts
@@ -553,8 +553,15 @@ export const acpStore = {
           "[AcpStore] Session appears dead, attempting auto-recovery...",
         );
 
-        // Preserve conversation history and cwd before cleanup
-        const existingMessages = [...session.messages];
+        // Preserve conversation history and cwd before cleanup.
+        // Filter out any "unresponsive" error messages that the event handler
+        // may have added before this catch block ran — restoring them would
+        // create duplicate banners in the new session.
+        const existingMessages = [...session.messages].filter(
+          (m) =>
+            m.id !== userMessage.id &&
+            !(m.type === "error" && m.content.includes("unresponsive")),
+        );
         const cwd = session.cwd;
         const agentType = session.info.agentType;
 
@@ -564,25 +571,30 @@ export const acpStore = {
         // Spawn a fresh session
         const newSessionId = await this.spawnSession(cwd, agentType);
         if (newSessionId) {
-          // Restore conversation history to the new session (excluding the
-          // user message we just added, since we'll retry the prompt)
-          const historyToRestore = existingMessages.filter(
-            (m) => m.id !== userMessage.id,
-          );
-          if (historyToRestore.length > 0) {
-            setState("sessions", newSessionId, "messages", historyToRestore);
+          // Restore conversation history to the new session
+          if (existingMessages.length > 0) {
+            setState("sessions", newSessionId, "messages", existingMessages);
           }
+
+          // Show recovery indicator so the user knows what happened
+          const recoveryMsg: AgentMessage = {
+            id: crypto.randomUUID(),
+            type: "assistant",
+            content:
+              "Agent session restarted due to inactivity timeout. Retrying your message...",
+            timestamp: Date.now(),
+          };
+          setState("sessions", newSessionId, "messages", (msgs) => [
+            ...msgs,
+            recoveryMsg,
+            userMessage,
+          ]);
 
           // Retry the prompt on the new session
           console.info(
             `[AcpStore] Retrying prompt on new session ${newSessionId}`,
           );
           try {
-            // Add the user message to the new session
-            setState("sessions", newSessionId, "messages", (msgs) => [
-              ...msgs,
-              userMessage,
-            ]);
             await acpService.sendPrompt(newSessionId, prompt, context);
             console.log("[AcpStore] Retry succeeded on new session");
             return;
@@ -592,7 +604,10 @@ export const acpStore = {
               retryError instanceof Error
                 ? retryError.message
                 : String(retryError);
-            this.addErrorMessage(newSessionId, retryMessage);
+            this.addErrorMessage(
+              newSessionId,
+              `Recovery failed: ${retryMessage}. Please try sending your message again.`,
+            );
             return;
           }
         }
@@ -875,6 +890,14 @@ export const acpStore = {
             ...msgs,
             cancelMsg,
           ]);
+        } else if (String(event.data.error).includes("unresponsive")) {
+          // "Agent unresponsive" errors are handled by the sendPrompt catch
+          // block which spawns a fresh session and retries. Adding the error
+          // here would create duplicate banners when the recovery code
+          // restores message history to the new session.
+          console.info(
+            "[AcpStore] Skipping error message for unresponsive agent — sendPrompt handles recovery",
+          );
         } else {
           this.addErrorMessage(sessionId, event.data.error);
         }

